theano/misc/pkl_utils.py

```diff
@@ -275,25 +275,31 @@ class PersistentNdarrayLoad(object):
     """
     def __init__(self, zip_file):
         self.zip_file = zip_file
+        self.cache = {}
 
     def __call__(self, persid):
         array_type, name = persid.split('.')
 
+        if name in self.cache:
+            return self.cache[name]
```

Comment:
why not use persid as the key of the cache? I think it is more safe. I think you can have those 2 persid that would conflict with the current code:

ndarray.array_0
cuda_ndarray.array_0

Thanks for investigating and founding the root of the problem. I suppose it wasn't trivial to find.